Per directory configs - preliminary changes

0nf opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [ ] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [ ] Write comprehensive commit messages and/or a good description of what the PR does.
- [ ] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes


|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Modifications in existing code base that are needed for per-directory configs. This PR does not introduce new functionality itself, but contains part of the changes from https://github.com/pylint-dev/pylint/pull/9395.

The only new behavior from this PR is slightly modified messages in verbose mode.

Refs #618

Commit:
Prepare code for per-directory configuration files

- Add opportunity to open checkers per-file, so they can use values from local config during opening
- Save command line arguments to apply them on top of each new config
- More accurate verbose messages about config files
- Enable finding config files in arbitrary directories
- Add opportunity to call linter._astroid_module_checker per file in single-process mode
- Collect stats from several calls of linter._astroid_module_checker in single-process mode
- Extend linter._get_namespace_for_file to return the path from which namespace was created

DanielNoord commented on requirements_test.txt:
```diff
@@ -9,3 +9,4 @@ six
 # Type packages for mypy
 types-pkg_resources==0.1.3
 tox>=3
+pre-commit
```
Can we leave this out? `pre-commit` is not necessary to run the tests.

DanielNoord replied:
I don't really use `tox`, didn't know we still recommend it.

You can probably just use the CI for this? That's what I always do :smile:

0nf replied:
Moved `pre-commit` dependency directly to the tox.ini config - does it seem like better solution?

DanielNoord commented on pylint/config/config_initialization.py:
```diff
@@ -82,6 +83,9 @@ def _config_initialization(
     args_list = _order_all_first(args_list, joined=True)
     parsed_args_list = linter._parse_command_line_configuration(args_list)
 
+    # save Runner.verbose to make this preprocessed option visible from other modules
+    linter.config.verbose = verbose_mode
```
I don't understand why this is necessary. Is ore logic for this flawed?

0nf replied:
Currently the only places where it is known that `--verbose` was set is inside `Run.__init__()` and in 2 functions that are called directly from there (`_config_initialization` and `linter.generate_reports`). 

When I add additional configs, I also add verbose messages about loading and applying of this configs. Since applying per-file configs can't happen in global `Run.__init__()`, information about verbosity should be copied to some place where several calls of `linter._lint_file` and downstream methods can access it.

DanielNoord commented on pylint/config/config_initialization.py:
```diff
@@ -141,7 +145,8 @@ def _config_initialization(
 
     # Link the base Namespace object on the current directory
-    linter._directory_namespaces[Path(".").resolve()] = (linter.config, {})
+    if Path(".").resolve() not in linter._directory_namespaces:
+        linter._directory_namespaces[Path(".").resolve()] = (linter.config, {})
```
This should not have the `if` I think? `_config_initialization` should be called once? Even for multi-dir?

0nf replied:
`_config_initialization` has some non-trivial logic for parsing several possible variants of configs into namespace, merging it with command-line arguments, configuring plugins and reporting errors during this process. So it was convenient to reuse all this logic for parsing additional configs, and `_config_initialization` is going to be called for each new config in subsequent changes

DanielNoord replied:
Hmm but shouldn't we pass the path of the current config file to this function then? On its own this if statement doesn't make a lot of sense.

0nf replied:
Exactly - the path of the current config file is passed to `_config_initialization` in `Run.__init__`, and paths of new config files will be passed there in `register_local_config`.

The idea of this `if` is that first time when `_config_initialization` is called, it parses config from working directory, and this config should be saved to `linter._directory_namespaces` to avoid additional processing of special cases. But next times when `_config_initialization` is called, it shouldn't overwrite config for working directory with values from new files.

DanielNoord commented on pylint/config/find_default_config_files.py:
```diff
@@ -64,17 +64,19 @@ def _cfg_has_config(path: Path | str) -> bool:
 
 
-def _yield_default_files() -> Iterator[Path]:
+def _yield_default_files(basedir: Path | str = ".") -> Iterator[Path]:
```
```suggestion
def _yield_default_files(basedir: Path  = Path(".")) -> Iterator[Path]:
```

?

0nf replied:
+

DanielNoord commented on pylint/config/find_default_config_files.py:
```diff
@@ -142,3 +144,8 @@ def find_default_config_files() -> Iterator[Path]:
         pass
+
+
+def find_subdirectory_config_files(basedir: Path | str) -> Iterator[Path]:
+    """Find config file in arbitrary subdirectory."""
+    yield from _yield_default_files(basedir)
```
This is not called in this PR so should probably go in a follow up

0nf replied:
done

DanielNoord commented on pylint/lint/pylinter.py:
```diff
@@ -66,7 +66,7 @@
     Options,
 )
-from pylint.utils import ASTWalker, FileState, LinterStats, utils
+from pylint.utils import ASTWalker, FileState, LinterStats, merge_stats, utils
```
Could you explain why we need the merging of stats for the multi-dir config option?

0nf replied:
Current state is that some stat counters are reset during `linter.open()`, some are reset in `linter.set_current_module` -> `init_single_module()`, and some
 are not reset at all (error, warning etc in 1st group, all counters in stats.by_module in 2nd group, statement in 3rd). It leads to incorrect score calculation when linter (i.e. main checker) is opened per file, or when it is opened after getting asts.
So I decided to reset all possible counters for each new module by creating new LinterStats object in `set_current_module`.

If stats reset is omitted entirely, then another problem arises:
When jobs>1, the same linter object can be used for checking several modules, stats after each module are copied and then merged.
It leads to a situation when some stats are accounted several times in final result (it's checked in `test_subconfigs_score` in my 1st PR).

Explicit stats reset and merge in single process can be avoided, but it will require additional changes in code for parallel checks. I'd suggest to leave it as a possible optimization in another PR.

DanielNoord replied:
Sorry, you probably already explain it but I don't fully understand. This explanation seems to point to a general issue with stats merging, not something that has to do with multi-directory configs. Or am I misunderstanding you? If it is just a general issue we should tackle it in a separate PR.

DanielNoord reviewed (commented):
Thanks! This makes it much easier to review. Left some comments!

Commit:
Fix formatting check when running tox

art049 commented:
Hey @0nf and @DanielNoord, I've been running the benches on this branch, and these changes seem to significantly impact the baseline benchmarks and from their definition, those seem important:
https://github.com/pylint-dev/pylint/blob/67bfab41b6a830cbab669b4eb15204690c64846d/tests/benchmark/test_baseline_benchmarks.py#L118-L124

However, there is a big regression on the runs:
![image](https://github.com/pylint-dev/pylint/assets/15641577/d9e14afa-692b-4f3d-8f16-90dad5a2f0e0)

Curious to know if you expected this performance change.
For some explanation, I installed CodSpeed on [a fork synced with this repo](https://github.com/CodSpeedHQ/pylint/pull/5). You can look at the full report [here](https://codspeed.io/CodSpeedHQ/pylint/branches/per_directory_configs_preliminary).

DanielNoord commented:
Thanks for that @art049.

It is probably related to the moving of `ast_per_fileitem = self._get_asts(fileitems, data)`. Does the report also point to what is making the performance slower?

art049 commented:
@DanielNoord yes it seems from the differential profile the regressions is mainly located in `PyLinter._get_namespace_for_file`:
<img width="682" alt="image" src="https://github.com/pylint-dev/pylint/assets/15641577/3c0d970b-f9af-4068-bae5-1456780a8272">

A lot of new code(in blue) is executed here.

0nf commented:
`Path.resolve()` is not connected to the moving of `self._get_asts(fileitems, data)`, it is just added for correct identification of parent directories, including situations where paths contain symlinks.

There is also a [report based on branch with full changes for per-directory configs](https://codspeed.io/CodSpeedHQ/pylint/branches/per_directory_configs). `_get_namespace_for_file` is behind the new feature flag there, so in the end performance of `test_baseline_benchmark_j1` is affected to a much lesser extent.

Commit:
Preliminary changes 2

- Responses to review comments
- Add test for calling _astroid_module_checker on different levels
- Move Path.resolve() out of _get_namespace_for_file recursive calls

DanielNoord commented:
@jacobtylerwalls I think you have done some regression testing in the past. Can you comment on whether you see a performance regression with these changes/

jacobtylerwalls commented on pylint/lint/pylinter.py:
```diff
@@ -898,33 +911,43 @@ def _expand_files(
         if filepath:
-            namespace = self._get_namespace_for_file(
-                Path(filepath), self._directory_namespaces
+            config_path, namespace = self._get_namespace_for_file(
+                Path(filepath).resolve(), self._directory_namespaces
```
The linked report traces the regression to the call to `resolve()`. 

I wonder if we can guard it under `not path.is_absolute()`:
```py
>>> from timeit import timeit
>>> timeit('p.is_absolute()', setup='from pathlib import Path; p=Path(".")')
0.1643185840221122
>>> timeit('p.resolve()', setup='from pathlib import Path; p=Path(".")')
10.929103292000946
```

jacobtylerwalls replied:
Edit: Seems like is_absolute() will always be false as things stand, so we probably need to look higher up the stack for a place to do some sort of conversion.

0nf commented:
Hi @DanielNoord ! I haven't marked some of your review notes as resolved because I don't know if my answers to them were sufficient. Could you check if my comments actually answer your questions? 🙂

0nf commented:
Also, I'm a bit confused about what to do with performance drop in test_baseline_benchmark_j1.
- Is it critical, given that the time difference is less than 5ms in all cases, which is <1% in all `test_baseline_lots_of_files*` benchmarks?
- If yes - would it be ok to hide `Path.resolve()` behind an analog of `use-local-configs` feature flag? I was thinking about a condition like `len(self._directory_namespaces) > 1`

DanielNoord commented:
Just letting you know that this is on my TODO list but just haven't found the time yet.

DanielNoord commented on pylint/config/config_initialization.py:
```diff
@@ -141,7 +145,8 @@ def _config_initialization(
 
     # Link the base Namespace object on the current directory
-    linter._directory_namespaces[Path(".").resolve()] = (linter.config, {})
+    if len(linter._directory_namespaces) == 0:
+        linter._directory_namespaces[Path(".").resolve()] = (linter.config, {})
```
For now I'd prefer to revert the changes in these two lines.

They are really rightly coupled to the final implementation of the per directory configs and the performance impact is hard to judge on its own. As far as I can see, all other changes in this PR are somewhat sensible on their own. This one isn't.

DanielNoord commented on pylint/config/find_default_config_files.py:
```diff
@@ -64,17 +64,19 @@ def _cfg_has_config(path: Path | str) -> bool:
-def _yield_default_files() -> Iterator[Path]:
+def _yield_default_files(basedir: Path = Path(".")) -> Iterator[Path]:
     """Iterate over the default config file names and see if they exist."""
+    basedir = Path(basedir)
```
```suggestion
```

That should not be needed.

DanielNoord commented on pylint/lint/pylinter.py:
```diff
@@ -693,13 +694,12 @@ def check(self, files_or_modules: Sequence[str]) -> None:
         with augmented_sys_path(extra_packages_paths):
+            # 2) Get the AST for each FileItem
+            ast_per_fileitem = self._get_asts(fileitems, data)
+            # 3) Lint each ast
```
This line should be down below

DanielNoord commented on pylint/lint/pylinter.py:
```diff
@@ -693,13 +694,12 @@ def check(self, files_or_modules: Sequence[str]) -> None:
-        # The contextmanager also opens all checkers and sets up the PyLinter class
         with augmented_sys_path(extra_packages_paths):
+            # 2) Get the AST for each FileItem
+            ast_per_fileitem = self._get_asts(fileitems, data)
```
I wonder if we have unintended effects from not getting these within the context manager..

DanielNoord commented on pylint/lint/pylinter.py:
```diff
@@ -760,12 +762,17 @@ def _lint_files(
         file: FileItem,
         module: nodes.Module,
-        check_astroid_module: Callable[[nodes.Module], bool | None],
+        check_astroid_module: Callable[[nodes.Module], bool | None] | None,
```
This PR is really touching some of the core behaviour of this behemoth of a class so it is a bit hard to review. Sorry in advance.

Why is this now optional? I don't really like that design as it further complicates this function body. Could you explain why this is needed? And could that perhaps be a separate PR?

DanielNoord reviewed (commented):
Thanks again for continuing with this @0nf

If you don't mind I could also split off some of the things I think we can easily merge into separate PRs to get them reviewed by other maintainers to make this PR a little bit more manageable.
